core/stateless/encoding.go

```diff
@@ -70,6 +146,105 @@ func (w *Witness) DecodeRLP(s *rlp.Stream) error {
 	return w.fromExtWitness(&ext)
 }
 
+// EncodeCompressed serializes a witness with optional compression.
+func (w *Witness) EncodeCompressed(wr io.Writer) error {
+	// First encode to RLP
+	var rlpBuf bytes.Buffer
+	if err := w.EncodeRLP(&rlpBuf); err != nil {
+		return err
+	}
+
+	rlpData := rlpBuf.Bytes()
+	originalSize := len(rlpData)
+
+	// Track original size
+	atomic.AddInt64(&totalOriginalSize, int64(originalSize))
+
+	// Only compress if enabled and the data is large enough to benefit from compression
+	if globalCompressionConfig.Enabled && len(rlpData) > globalCompressionConfig.Threshold {
+		// Compress the RLP data
+		var compressedBuf bytes.Buffer
+		gw, err := gzip.NewWriterLevel(&compressedBuf, globalCompressionConfig.CompressionLevel)
+		if err != nil {
+			return err
+		}
+
+		if _, err := gw.Write(rlpData); err != nil {
+			return err
+		}
+
+		if err := gw.Close(); err != nil {
+			return err
+		}
+
+		compressedData := compressedBuf.Bytes()
+
+		// Only use compression if it actually reduces size
+		if len(compressedData) < len(rlpData) {
+			// Track compression metrics
+			atomic.AddInt64(&compressionCount, 1)
+			atomic.AddInt64(&totalCompressedSize, int64(len(compressedData)))
+			ratio := int64(float64(len(compressedData)) / float64(originalSize) * 100)
+			atomic.AddInt64(&compressionRatio, ratio)
+
+			// Write compression marker and compressed data
+			if _, err := wr.Write([]byte{0x01}); err != nil {
+				return err
+			}
+			_, err = wr.Write(compressedData)
+			return err
+		}
+	}
+
+	// Track uncompressed metrics
+	atomic.AddInt64(&uncompressedCount, 1)
+	atomic.AddInt64(&totalCompressedSize, int64(originalSize))
+
+	// Write uncompressed marker and original RLP data
+	if _, err := wr.Write([]byte{0x00}); err != nil {
+		return err
+	}
+	_, err := wr.Write(rlpData)
+	return err
+}
+
+// DecodeCompressed decodes a witness from compressed format.
+func (w *Witness) DecodeCompressed(data []byte) error {
```

Comment:
Similarly, metrics for decompression would be useful.

Reply:
[Done](https://github.com/maticnetwork/bor/pull/1648/commits/c27eb8b9c2bce61a40e4ee8ca89cbf3e92f6c153).